build_tools/github_actions/ci_build_bazel.sh

```diff
@@ -16,11 +16,77 @@ set -o errexit
 set -o nounset
 set -o pipefail
 
-if [[ $# -ne 0 ]] ; then
-  echo "Usage: $0"
+if [[ $# -ne 0 && $# -ne 3 ]] ; then
+  echo "Usage: $0 [<bazel-diff.jar> <start_commit> <end_commit>]"
+  echo "   Builds and tests all bazel targets."
+  echo "   If start_commit and end_commit are specified, uses bazel-diff to"
+  echo "   determine the set of targets to build based on code changes."
+  echo
+  echo "Example: Build all affected targets between current branch and main"
+  echo "  $ curl -Lo /tmp/bazel-diff.jar https://github.com/Tinder/bazel-diff/releases/latest/download/bazel-diff_deploy.jar"
+  echo "  $ ci_build_bazel.sh /tmp/bazel-diff.jar main $(git rev-parse --abbrev-ref HEAD)"
   exit 1
 fi
 
-# Build and Test StableHLO
-bazel build --lockfile_mode=error //... --config=asan --config=ubsan
-bazel test //... --config=asan --config=ubsan
+bazel-test-all() {
+  # Build and Test StableHLO
+  bazel build --lockfile_mode=error //... --config=asan --config=ubsan
+  bazel test //... --config=asan --config=ubsan
+}
+
+bazel-test-diff() {
+  set -e
+  WORKSPACE_PATH=$(git rev-parse --show-toplevel)
+  BAZEL_DIFF_JAR="$1"
+  PREVIOUS_REV="$2" # Starting Revision SHA
+  FINAL_REV="$3" # Final Revision SHA
+  BAZEL_PATH="$(which bazel)"
+
+  STARTING_HASHES_JSON="/tmp/starting_hashes.json"
+  FINAL_HASHES_JSON="/tmp/final_hashes.json"
+  IMPACTED_TARGETS_PATH="/tmp/impacted_targets.txt"
+
+  bazel-diff() {
+    java -jar "$BAZEL_DIFF_JAR" "$@"
+  }
+
+  git -C "$WORKSPACE_PATH" checkout "$PREVIOUS_REV" --quiet
+
+  echo "Generating Hashes for Revision '$PREVIOUS_REV'"
+  bazel-diff generate-hashes -w "$WORKSPACE_PATH" -b "$BAZEL_PATH" $STARTING_HASHES_JSON
+
+  git -C "$WORKSPACE_PATH" checkout "$FINAL_REV" --quiet
+
+  echo "Generating Hashes for Revision '$FINAL_REV'"
+  bazel-diff generate-hashes -w "$WORKSPACE_PATH" -b "$BAZEL_PATH" $FINAL_HASHES_JSON
+
+  echo "Determining Impacted Targets"
+  bazel-diff get-impacted-targets -sh $STARTING_HASHES_JSON -fh $FINAL_HASHES_JSON -o $IMPACTED_TARGETS_PATH
+  echo ""
+
+  impacted_targets=()
+  IFS=$'\n' read -d '' -r -a impacted_targets < $IMPACTED_TARGETS_PATH || true
```

Comment:
How about we save/restore IFS 
```
IFS_OLD="$IFS"
IFS= .... || true
IFS="$IFS_OLD"
```

to keep the bash string splitting behavior predictable for any future modification of this file.

Reply:
This is copy-pate-edited from the `bazel-diff` upstream script, so probably will keep as is to keep any future syncing of this file easier